### DISCUSSION TITLE
Is there an SPI for a custom exporter?

### DISCUSSION THREAD

**Discussion by daschneider** [open-telemetry/opentelemetry-java#4722] [Q&A]
We'd like to export all metrics to a local file while developers are working on their laptop for testing and troubleshooting purposes.  Is there an SPI available where we can inject our own customer exporter?

**Comment by mateuszrzeszutek** [comment]
Hey @daschneider ,
You can try using the [`ConfigurableMetricExporterProvider`](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/metrics/ConfigurableMetricExporterProvider.java) SPI to provide a metric exporter implementation -- you'll have to set the `otel.metrics.exporter` configuration setting to your chosen exporter name.

**Reply by daschneider** [reply]
Thanks @mateuszrzeszutek - this looks like exactly what I was looking for.  Trying it out now.

**Reply by daschneider** [reply]
@mateuszrzeszutek I'm having trouble getting the SPI to load.  I'm using the 1.17.0 java agent and the following SDK version:
 ```
    implementation platform("io.opentelemetry:opentelemetry-bom:1.17.0")
    implementation platform('io.opentelemetry:opentelemetry-bom-alpha:1.17.0-alpha')

    implementation('io.opentelemetry:opentelemetry-api')
    implementation('io.opentelemetry:opentelemetry-sdk-extension-autoconfigure')
```
Are these the correct versions?

I have a `META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.metrics.ConfigurableMetricExporterProvider` file in my classpath and it contains the name of my class implementing the `ConfigurableMetricExporterProvider` interface, but my implementation class is never instantiated.

Does all of this look correct?  Any idea why my provider class is not being loaded?

**Reply by trask** [reply]
hi @daschneider, since you are using the java agent, you'll need to package your exporter in an java agent extension, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension
